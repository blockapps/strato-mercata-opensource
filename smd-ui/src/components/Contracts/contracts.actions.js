export const FETCH_CONTRACTS = 'FETCH_CONTRACTS';
export const FETCH_CONTRACTS_SUCCESSFUL = 'FETCH_CONTRACTS_SUCCESSFUL';
export const FETCH_CONTRACTS_FAILED = 'FETCH_CONTRACTS_FAILED';
export const CHANGE_CONTRACT_FILTER = 'CHANGE_CONTRACT_FILTER';
export const TOGGLE_COLLAPSE_TABLE = 'TOGGLE_COLLAPSE_TABLE';

export const fetchContracts = function (chainId, limit, offset, name) {
  return {
    type: FETCH_CONTRACTS,
    chainId,
    limit,
    offset,
    name
  }
};

export const fetchContractsSuccess = function (contracts) {
  return {
    type: FETCH_CONTRACTS_SUCCESSFUL,
    contracts: contracts
  }
};

export const fetchContractsFailure = function (error) {
  return {
    type: FETCH_CONTRACTS_FAILED,
    error: error,
  }
};

export const changeContractFilter = function (filter) {
  return {
    type: CHANGE_CONTRACT_FILTER,
    filter: filter,
  }
};
