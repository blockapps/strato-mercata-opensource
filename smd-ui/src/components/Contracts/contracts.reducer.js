import {
  FETCH_CONTRACTS,
  FETCH_CONTRACTS_SUCCESSFUL,
  FETCH_CONTRACTS_FAILED,
  CHANGE_CONTRACT_FILTER,
} from './contracts.actions';
import {
  FETCH_STATE_SUCCESS,
  FETCH_CIRRUS_INSTANCES_SUCCESS,
  SELECT_CONTRACT_INSTANCE,
  FETCH_ACCOUNT_SUCCESS
} from './components/ContractCard/contractCard.actions';

const initialState = {
  contracts: {},
  filter: '',
  error: null,
  isLoading: false
};

const reducer = function (state = initialState, action) {
  switch (action.type) {
    case FETCH_CONTRACTS:
      return {
        contracts: state.contracts,
        filter: state.filter,
        error: null,
        isLoading: true
      };
    case FETCH_ACCOUNT_SUCCESS:
      let contracts = state.contracts;
      const contractArray = state.contracts[action.name].instances;
      contractArray.forEach((contract) => {
        if (action.address === contract.address) {
          contract['balance'] = action.account.length > 0 ? action.account[0].balance : "0";
        }
      })
      contracts[action.name].instances = contractArray;
      return {
        contracts,
        filter: state.filter,
        error: null
      };
    case FETCH_CONTRACTS_SUCCESSFUL:
      const contractNames = Object.getOwnPropertyNames(action.contracts);
      const updatedContracts = {};
      contractNames.forEach((name) => {
        if (state.contracts[name]) {
          // new instances
          const newInstances = action.contracts[name]
            .filter((instance) => {
              return state.contracts[name].instances
                .filter((i) => {
                  return i.address === instance.address
                }).length === 0;
            })
            .map((instance) => {
              return {
                ...instance,
                fromBloc: true
              }
            });
          updatedContracts[name] = {
            instances: state.contracts[name].instances.concat(newInstances)
          };
        }
        else {
          updatedContracts[name] = {
            instances: action.contracts[name]
              .map((instance) => {
                return {
                  ...instance,
                  fromBloc: true
                }
              })
          };
        }
      });
      return {
        contracts: updatedContracts,
        filter: state.filter,
        error: state.error,
        isLoading: false
      };
    case FETCH_CONTRACTS_FAILED:
      return {
        contracts: state.contracts,
        filter: state.filter,
        error: action.error,
        isLoading: false
      };
    case CHANGE_CONTRACT_FILTER:
      return {
        contracts: state.contracts,
        filter: action.filter,
        error: state.error
      }
    case FETCH_STATE_SUCCESS:
      const instances = state.contracts[action.name].instances
        .map(function (instance, i) {
          if (instance.address !== action.address) {
            return instance
          }
          return {
            ...instance,
            state: action.state
          }
        });

      return {
        contracts: {
          ...state.contracts,
          [action.name]: {
            instances: instances
          }
        },
        filter: state.filter,
        error: state.error
      }
    case FETCH_CIRRUS_INSTANCES_SUCCESS:
      const updatedInstances = state.contracts[action.name].instances.map((instance) => {
        return {
          ...instance,
          fromCirrus: action.instances.filter((i) => {
            return i.address === instance.address
          }).length === 1
        };
      });

      action.instances.forEach((instance) => {
        const exists = updatedInstances.filter((i) => {
          return i.address === instance.address
        }).length === 1;
        if (!exists) {
          updatedInstances.push({
            ...instance,
            fromCirrus: true
          });
        }
      });
      return {
        contracts: {
          ...state.contracts,
          [action.name]: {
            instances: updatedInstances
          }
        },
        filter: state.filter,
        error: state.error
      };
    case SELECT_CONTRACT_INSTANCE:
      const cInstances = state.contracts[action.name].instances
        .map(function (instance, i) {
          return {
            ...instance,
            selected: instance.address === action.address
          }
        });
      return {
        contracts: {
          ...state.contracts,
          [action.name]: {
            instances: cInstances
          }
        },
        filter: state.filter,
        error: state.error
      }

    default:
      return state;
  }
};

export default reducer;
