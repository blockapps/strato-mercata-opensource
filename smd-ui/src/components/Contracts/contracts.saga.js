import {
  takeEvery,
  put,
  call
} from 'redux-saga/effects';
import {
  FETCH_CONTRACTS,
  fetchContractsSuccess,
  fetchContractsFailure
} from './contracts.actions';
import { env } from '../../env';
import { handleErrors } from '../../lib/handleErrors';

const contractsUrl = env.BLOC_URL + "/contracts";

export function getContracts(chainid, limit, offset, name) {
  const url = `${contractsUrl}?limit=${limit}&offset=${offset}${chainid ? `&chainid=${chainid}` : ''}${name ? `&name=${name}` : ''}`

  return fetch(
    url,
    {
      method: 'GET',
      credentials: "include",
      headers: {
        'Accept': 'application/json'
      },
    })
    .then(handleErrors)
    .then(function (response) {
      return response.json()
    })
    .catch(function (error) {
      throw error;
    });
}

export function* fetchContracts(action) {
  try {
    let response = yield call(getContracts, action.chainId, action.limit, action.offset, action.name);
    yield put(fetchContractsSuccess(response));
  }
  catch (err) {
    yield put(fetchContractsFailure(err));
  }
}

export default function* watchFetchContracts() {
  yield takeEvery(FETCH_CONTRACTS, fetchContracts);
}
