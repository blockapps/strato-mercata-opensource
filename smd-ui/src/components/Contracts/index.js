import React, { Component } from 'react';
import { fetchContracts, changeContractFilter } from './contracts.actions';
import { connect } from 'react-redux';
import CreateContract from '../CreateContract';
import ContractCard from './components/ContractCard';
import mixpanelWrapper from '../../lib/mixpanelWrapper';
import Tour from '../Tour';
import { Button, Popover, PopoverInteractionKind, Position, Switch, Tooltip} from '@blueprintjs/core';
import ReactGA from 'react-ga4';
import { Field, reduxForm } from 'redux-form';
import { selectChain, fetchChainIds, fetchChainDetailSelect } from '../Chains/chains.actions';
import { withRouter } from 'react-router-dom';
import HexText from '../HexText';

const tourSteps = [
  /*  {
      title: 'Create Contract',
      text: 'Where queries are to be found.',
      selector: '#tour-create-contract-button',
      position: 'bottom', type: 'hover',
      isFixed: true,
    }, */
  {
    title: 'View Transactions',
    text: 'Scroll through all transactions launched from your Smart Contract.',
    selector: '#transactions',
    position: 'bottom',
    isFixed: true,
  },
];

class Contracts extends Component {
  constructor(props) {
    super(props);
    this.state = {
      limit: 10,
      offset: 0,
      chainLimit: 25,
      chainOffset: 0,
      useSearch: true,
      chainSearchQueryField: "chainid",
      chainQuery: "",
    }
  }

  componentWillMount() {
    mixpanelWrapper.track("contracts_loaded");
    this.props.changeContractFilter('');
    this.props.fetchContracts(this.props.selectedChain, this.state.limit, this.state.offset);
    this.props.fetchChainIds(this.chainLimit, this.chainOffset);
  }

  componentDidMount() {
    ReactGA.send({hitType: "pageview", page: "/contracts", title: "Contracts"});
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.selectedChain !== this.props.selectedChain) {
      this.props.fetchContracts(nextProps.selectedChain, this.state.limit, this.state.offset, this.props.filter);
    }
  }

  updateFilter = (filter) => {
    this.props.changeContractFilter(filter);
    this.props.fetchContracts(this.props.selectedChain, this.state.limit, this.state.offset, filter);
  }

  onChainSearch = () => {
    this.props.fetchChainDetailSelect(this.state.chainQuery, this.state.chainSearchQueryField)
  }

  onNextClick = () => {
    const { offset, limit } = this.state;
    const newOffset = offset + limit;
    this.setState({ offset: newOffset }, () => {
      this.props.fetchContracts(this.props.selectedChain, this.state.limit, this.state.offset, this.props.filter);
    });
  };

  onPrevClick = () => {
    const { offset, limit } = this.state;
    const newOffset = Math.max(0, offset - limit);
    this.setState({ offset: newOffset }, () => {
      this.props.fetchContracts(this.props.selectedChain, this.state.limit, this.state.offset, this.props.filter);
    });
  };

  onNextChainClick = () => {
    const { chainOffset, chainLimit } = this.state;
    const newOffset = chainOffset + chainLimit;
    this.setState({ chainOffset: newOffset }, () => {
      this.props.fetchChainIds(this.state.chainLimit, this.state.chainOffset);
    });
  };

  onPrevChainClick = () => {
    const { chainOffset, chainLimit } = this.state;
    const newOffset = Math.max(0, chainOffset - chainLimit);
    this.setState({ chainOffset: newOffset }, () => {
      this.props.fetchChainIds(this.state.chainLimit, this.state.chainOffset);
    });
  };

  toggleChainQueryType = (e) => {
    this.setState({ useSearch : !this.state.useSearch }, () => {
      if (!this.state.useSearch) {
        this.setState({chainQuery : ""})
        this.props.fetchChainIds(this.chainLimit, this.chainOffset);
      }

    })
  }

  render() {
    const contracts = this.props.contracts;
    const contractNames = Object.getOwnPropertyNames(this.props.contracts);

    const returnedInstances = contracts && contractNames.length > 0 ? Object.values(contracts).reduce((prev, cur) => {
      return prev + cur.instances.length
    }, 0) : 0

    const cards = contractNames.length === 0 ? [] : contractNames
      .map((value, i) => {
        return (
          <div className="row pt-dark" key={'contract-card-' + i}>
            <div className="col-sm-12">
              {value && <ContractCard contract={{ name: value, contract: contracts[value] }} />}
              <br />
            </div>
          </div>
        );
      });
    const isPaginationDisplay = cards.length ? true : Boolean(this.state.offset);
      return (
      <div className="container-fluid">
        <Tour steps={tourSteps} name="contracts" finalStepSelector='#transactions' nextPage='transactions' />
        <div className="row pt-dark" >
          <div className="col-sm-2 text-left">
            <h3>Contracts</h3>
          </div>
        <div className="col-sm-6 smd-pad-16">
            <div className="pt-input-group pt-dark pt-large">
              <span className="pt-icon pt-icon-search"></span>
              <input
                className="pt-input"
                type="search"
                placeholder="Search contracts"
                onChange={e => this.updateFilter(e.target.value)}
                dir="auto" />
            </div>
          </div>
    
        </div>
        <div className='row pt-dark' style={{ display: 'flex', alignItems: 'center'}}>
            <div className="col-sm-2 text-right smd-pad-8">
              <CreateContract />
            </div>
        </div>
        {!cards.length && !this.props.isLoading &&
          <div className="row pt-dark" key={'contract-card-'}>
            <div className="col-sm-12">
              <h4>No Contracts</h4>
              <h5>Upload a Contract to View State</h5>
              <br />
            </div>
          </div>}
        {this.props.isLoading ?
          <div className="row pt-dark">
            <div className="col-sm-12">
              <div className="row">
                <div className="col-sm-6">
                  <div className="pt-card pt-elevation-2">
                    <div className="row">
                      <div className="col-sm-4"><h4>Working...</h4></div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          : cards
        }
        {isPaginationDisplay && <div className="pt-dark">
          <div className="row">
            <div className="col-sm-2 smd-pad-16 text-left">
              <Button
                onClick={this.onPrevClick}
                className="pt-icon-arrow-left"
                text="Previous"
                disabled={!(this.state.offset > 0)}
              />
            </div>
            <div className="col-sm-2 text-center" style={{ marginTop: '22px' }}>
              {`Rows ${this.state.offset + 1}-${this.state.offset + Math.min(cards.length, this.state.limit)} (${returnedInstances} Contract Instances)`}
            </div>
            <div className="col-sm-2 smd-pad-16 text-right">
              <Button
                onClick={this.onNextClick}
                className="pt-icon-arrow-right"
                text="Next"
                disabled={returnedInstances < this.state.limit}
              />
            </div>
          </div>
        </div>}
      </div>
    );
  }
}

export function mapStateToProps(state) {
  return {
    contracts: state.contracts.contracts,
    filter: state.contracts.filter,
    selectedChain: state.chains.selectedChain,
    isLoading: state.contracts.isLoading,
    chainIds: state.chains.chainIds,
  };
}

const formed = reduxForm({ form: 'Contracts' })(Contracts);
const connected = connect(mapStateToProps, {
  selectChain,
  fetchChainIds,
  fetchChainDetailSelect,
  fetchContracts, 
  changeContractFilter
})(formed);

export default withRouter(connected);
