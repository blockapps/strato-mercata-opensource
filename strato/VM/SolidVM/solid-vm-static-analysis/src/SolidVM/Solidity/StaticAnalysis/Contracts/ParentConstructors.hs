{-# LANGUAGE LambdaCase #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE RecordWildCards #-}

module SolidVM.Solidity.StaticAnalysis.Contracts.ParentConstructors
  ( detector,
  )
where

import Data.Functor ((<&>))
import qualified Data.Map.Strict as M
import Data.Maybe (fromMaybe)
import qualified Data.Set as S
import Data.Source
import Data.Text (Text)
import SolidVM.Model.CodeCollection
import SolidVM.Model.SolidString
import SolidVM.Solidity.StaticAnalysis.Types

-- type CompilerDetector = CodeCollection -> [SourceAnnotation T.Text]
detector :: CompilerDetector
detector cc@CodeCollection {..} = M.foldMapWithKey (contractHelper cc) _contracts

contractHelper :: CodeCollection -> SolidString -> Contract -> [SourceAnnotation Text]
contractHelper cc cName c =
  fromMaybe [] $
    _constructor c <&> \constr ->
      let parentSet = S.fromList $ _parents c
          constrCalls = _funcConstructorCalls constr
       in flip M.foldMapWithKey constrCalls $ \parentName varExprs ->
            if not $ parentName `S.member` parentSet
              then [("Contract " <> labelToText cName <> " does not inherit from " <> labelToText parentName) <$ _funcContext constr]
              else case M.lookup parentName (_contracts cc) of
                Nothing -> [("Contract " <> labelToText parentName <> " not found.") <$ _funcContext constr]
                Just parent -> case _constructor parent of
                  Nothing -> [(labelToText parentName <> "'s constructor is undefined. Please consider defining its constructor.") <$ _funcContext constr]
                  Just pConstr ->
                    let pConstrArgs = _funcArgs pConstr
                     in if length pConstrArgs /= length varExprs
                          then [("The number of arguments in the constructor call to " <> labelToText parentName <> " does not equal the number of arguments in its constructor definition.") <$ _funcContext constr]
                          else [] -- we'll leave typechecking the arguments for a different detector
