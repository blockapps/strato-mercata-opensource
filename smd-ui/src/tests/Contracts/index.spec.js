import React from 'react';
import Contracts, { mapStateToProps } from '../../components/Contracts/index';
import { chainIds, contracts } from './contractsMock';
import { Provider } from 'react-redux';
import { createStore, combineReducers } from 'redux';
import { reducer as formReducer } from 'redux-form';

describe('Contracts: index', () => {

  let store;

  beforeEach(() => {
    store = createStore(combineReducers({ form: formReducer }));
  });

  describe('render contracts with', () => {

    test('empty values', () => {
      const props = {
        filter: '',
        contracts: {},
        chainIds: [],
        fetchContracts: jest.fn(),
        changeContractFilter: jest.fn()
      }
      const wrapper = shallow(
        <Contracts.WrappedComponent {...props} />
      );
      expect(wrapper).toMatchSnapshot();
    });

    test('mocked values', () => {
      const props = {
        filter: 'Greeter',
        contracts: contracts,
        chainIds: chainIds,
        fetchContracts: jest.fn(),
        changeContractFilter: jest.fn()
      }
      const wrapper = shallow(
        <Contracts.WrappedComponent {...props} />
      );
      expect(wrapper).toMatchSnapshot();
    });

  })

  test('component methods', () => {
    const props = {
      filter: '',
      contracts: {},
      fetchContracts: jest.fn().mockReturnValue('fetchContracts'),
      changeContractFilter: jest.fn().mockReturnValue('changeContractFilter')
    }
    const wrapper = shallow(
      <Contracts.WrappedComponent {...props} />
    );
    wrapper.instance().updateFilter = jest.fn().mockReturnValue('updateFilter');
    expect(wrapper.instance().updateFilter()).toBe('updateFilter');
    expect(wrapper.instance().props.fetchContracts()).toBe('fetchContracts')
    expect(wrapper.instance().props.changeContractFilter()).toBe('changeContractFilter');
  });

  test('filter with value', () => {
    const props = {
      filter: 'Greeter',
      contracts: {},
      fetchContracts: jest.fn(),
      changeContractFilter: jest.fn()
    }
    const wrapper = shallow(
      <Contracts.WrappedComponent {...props} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  test('contracts with values', () => {
    const props = {
      filter: '',
      contracts: contracts,
      fetchContracts: jest.fn(),
      changeContractFilter: jest.fn()
    }
    const wrapper = shallow(
      <Contracts.WrappedComponent {...props} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  test('search contract input', () => {
    const props = {
      filter: 'Greeter',
      contracts: {},
      chainIds: [],
      store: store,
      fetchContracts: jest.fn(),
      changeContractFilter: jest.fn(),
      fetchChainIds: jest.fn()
    }
    const wrapper = shallow(
      <Provider store={store}>
        <Contracts.WrappedComponent {...props} />
      </Provider>
    ).dive().dive().dive().dive();

    wrapper.find('input').at(0).simulate('change', { target: { value: "UPDATE" } });
    expect(props.changeContractFilter).toHaveBeenCalled();
    expect(props.changeContractFilter.mock.calls.length).toBe(2);
    expect(props.changeContractFilter.mock.calls).toMatchSnapshot();
  });

  test('mapStateToProps with default values', () => {
    const state = {
      contracts: {
        contracts: contracts,
        filter: 'Time'
      },
      chains: {
        selectedChain: "ff7ef45acb7a775018bc765b6fdeea432aaddfcd846cf6dd9442724266b1eac9",
        chainIds: chainIds
      },
    }
    expect(mapStateToProps(state)).toMatchSnapshot();
  });

})