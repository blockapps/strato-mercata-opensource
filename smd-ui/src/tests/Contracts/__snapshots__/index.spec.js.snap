// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contracts: index contracts with values 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ReduxForm
    changeContractFilter={[Function]}
    contracts={
      Object {
        "Cloner": Array [
          Object {
            "address": "d07e932212f7f368b6948ffd96e1d4c726c8395d",
            "createdAt": 1512480630000,
          },
          Object {
            "address": "2c6619f0418c2f191e2225091f7692363a91c336",
            "createdAt": 1512480770000,
          },
        ],
        "GreeterA": Array [
          Object {
            "address": "0293f9b10a4453667db7fcfe74728c9d821add4b",
            "createdAt": 1512481078000,
          },
        ],
        "GreeterB": Array [
          Object {
            "address": "GreeterA",
            "createdAt": 1512481078000,
          },
        ],
      }
    }
    fetchContracts={[Function]}
    filter=""
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "changeContractFilter": [Function],
      "contracts": Object {
        "Cloner": Array [
          Object {
            "address": "d07e932212f7f368b6948ffd96e1d4c726c8395d",
            "createdAt": 1512480630000,
          },
          Object {
            "address": "2c6619f0418c2f191e2225091f7692363a91c336",
            "createdAt": 1512480770000,
          },
        ],
        "GreeterA": Array [
          Object {
            "address": "0293f9b10a4453667db7fcfe74728c9d821add4b",
            "createdAt": 1512481078000,
          },
        ],
        "GreeterB": Array [
          Object {
            "address": "GreeterA",
            "createdAt": 1512481078000,
          },
        ],
      },
      "destroyOnUnmount": true,
      "enableReinitialize": false,
      "fetchContracts": [Function],
      "filter": "",
      "forceUnregisterOnUnmount": false,
      "form": "Contracts",
      "getFormState": [Function],
      "initialValues": undefined,
      "keepDirtyOnReinitialize": false,
      "persistentSubmitErrors": false,
      "pure": true,
      "shouldAsyncValidate": [Function],
      "shouldValidate": [Function],
      "touchOnBlur": true,
      "touchOnChange": false,
    },
    "ref": "wrapped",
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "changeContractFilter": [Function],
        "contracts": Object {
          "Cloner": Array [
            Object {
              "address": "d07e932212f7f368b6948ffd96e1d4c726c8395d",
              "createdAt": 1512480630000,
            },
            Object {
              "address": "2c6619f0418c2f191e2225091f7692363a91c336",
              "createdAt": 1512480770000,
            },
          ],
          "GreeterA": Array [
            Object {
              "address": "0293f9b10a4453667db7fcfe74728c9d821add4b",
              "createdAt": 1512481078000,
            },
          ],
          "GreeterB": Array [
            Object {
              "address": "GreeterA",
              "createdAt": 1512481078000,
            },
          ],
        },
        "destroyOnUnmount": true,
        "enableReinitialize": false,
        "fetchContracts": [Function],
        "filter": "",
        "forceUnregisterOnUnmount": false,
        "form": "Contracts",
        "getFormState": [Function],
        "initialValues": undefined,
        "keepDirtyOnReinitialize": false,
        "persistentSubmitErrors": false,
        "pure": true,
        "shouldAsyncValidate": [Function],
        "shouldValidate": [Function],
        "touchOnBlur": true,
        "touchOnChange": false,
      },
      "ref": "wrapped",
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenAdapter {
      "options": Object {
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
  },
}
`;

exports[`Contracts: index filter with value 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ReduxForm
    changeContractFilter={[Function]}
    contracts={Object {}}
    fetchContracts={[Function]}
    filter="Greeter"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "changeContractFilter": [Function],
      "contracts": Object {},
      "destroyOnUnmount": true,
      "enableReinitialize": false,
      "fetchContracts": [Function],
      "filter": "Greeter",
      "forceUnregisterOnUnmount": false,
      "form": "Contracts",
      "getFormState": [Function],
      "initialValues": undefined,
      "keepDirtyOnReinitialize": false,
      "persistentSubmitErrors": false,
      "pure": true,
      "shouldAsyncValidate": [Function],
      "shouldValidate": [Function],
      "touchOnBlur": true,
      "touchOnChange": false,
    },
    "ref": "wrapped",
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "changeContractFilter": [Function],
        "contracts": Object {},
        "destroyOnUnmount": true,
        "enableReinitialize": false,
        "fetchContracts": [Function],
        "filter": "Greeter",
        "forceUnregisterOnUnmount": false,
        "form": "Contracts",
        "getFormState": [Function],
        "initialValues": undefined,
        "keepDirtyOnReinitialize": false,
        "persistentSubmitErrors": false,
        "pure": true,
        "shouldAsyncValidate": [Function],
        "shouldValidate": [Function],
        "touchOnBlur": true,
        "touchOnChange": false,
      },
      "ref": "wrapped",
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenAdapter {
      "options": Object {
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
  },
}
`;

exports[`Contracts: index mapStateToProps with default values 1`] = `
Object {
  "chainIds": Array [
    Object {
      "id": "0293f9b10a4453667db7fcfe74728c9d821add4b",
      "label": "new",
    },
  ],
  "contracts": Object {
    "Cloner": Array [
      Object {
        "address": "d07e932212f7f368b6948ffd96e1d4c726c8395d",
        "createdAt": 1512480630000,
      },
      Object {
        "address": "2c6619f0418c2f191e2225091f7692363a91c336",
        "createdAt": 1512480770000,
      },
    ],
    "GreeterA": Array [
      Object {
        "address": "0293f9b10a4453667db7fcfe74728c9d821add4b",
        "createdAt": 1512481078000,
      },
    ],
    "GreeterB": Array [
      Object {
        "address": "GreeterA",
        "createdAt": 1512481078000,
      },
    ],
  },
  "filter": "Time",
  "isLoading": undefined,
  "selectedChain": "ff7ef45acb7a775018bc765b6fdeea432aaddfcd846cf6dd9442724266b1eac9",
}
`;

exports[`Contracts: index render contracts with empty values 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ReduxForm
    chainIds={Array []}
    changeContractFilter={[Function]}
    contracts={Object {}}
    fetchContracts={[Function]}
    filter=""
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "chainIds": Array [],
      "changeContractFilter": [Function],
      "contracts": Object {},
      "destroyOnUnmount": true,
      "enableReinitialize": false,
      "fetchContracts": [Function],
      "filter": "",
      "forceUnregisterOnUnmount": false,
      "form": "Contracts",
      "getFormState": [Function],
      "initialValues": undefined,
      "keepDirtyOnReinitialize": false,
      "persistentSubmitErrors": false,
      "pure": true,
      "shouldAsyncValidate": [Function],
      "shouldValidate": [Function],
      "touchOnBlur": true,
      "touchOnChange": false,
    },
    "ref": "wrapped",
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "chainIds": Array [],
        "changeContractFilter": [Function],
        "contracts": Object {},
        "destroyOnUnmount": true,
        "enableReinitialize": false,
        "fetchContracts": [Function],
        "filter": "",
        "forceUnregisterOnUnmount": false,
        "form": "Contracts",
        "getFormState": [Function],
        "initialValues": undefined,
        "keepDirtyOnReinitialize": false,
        "persistentSubmitErrors": false,
        "pure": true,
        "shouldAsyncValidate": [Function],
        "shouldValidate": [Function],
        "touchOnBlur": true,
        "touchOnChange": false,
      },
      "ref": "wrapped",
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenAdapter {
      "options": Object {
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
  },
}
`;

exports[`Contracts: index render contracts with mocked values 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ReduxForm
    chainIds={
      Array [
        Object {
          "id": "0293f9b10a4453667db7fcfe74728c9d821add4b",
          "label": "new",
        },
      ]
    }
    changeContractFilter={[Function]}
    contracts={
      Object {
        "Cloner": Array [
          Object {
            "address": "d07e932212f7f368b6948ffd96e1d4c726c8395d",
            "createdAt": 1512480630000,
          },
          Object {
            "address": "2c6619f0418c2f191e2225091f7692363a91c336",
            "createdAt": 1512480770000,
          },
        ],
        "GreeterA": Array [
          Object {
            "address": "0293f9b10a4453667db7fcfe74728c9d821add4b",
            "createdAt": 1512481078000,
          },
        ],
        "GreeterB": Array [
          Object {
            "address": "GreeterA",
            "createdAt": 1512481078000,
          },
        ],
      }
    }
    fetchContracts={[Function]}
    filter="Greeter"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "chainIds": Array [
        Object {
          "id": "0293f9b10a4453667db7fcfe74728c9d821add4b",
          "label": "new",
        },
      ],
      "changeContractFilter": [Function],
      "contracts": Object {
        "Cloner": Array [
          Object {
            "address": "d07e932212f7f368b6948ffd96e1d4c726c8395d",
            "createdAt": 1512480630000,
          },
          Object {
            "address": "2c6619f0418c2f191e2225091f7692363a91c336",
            "createdAt": 1512480770000,
          },
        ],
        "GreeterA": Array [
          Object {
            "address": "0293f9b10a4453667db7fcfe74728c9d821add4b",
            "createdAt": 1512481078000,
          },
        ],
        "GreeterB": Array [
          Object {
            "address": "GreeterA",
            "createdAt": 1512481078000,
          },
        ],
      },
      "destroyOnUnmount": true,
      "enableReinitialize": false,
      "fetchContracts": [Function],
      "filter": "Greeter",
      "forceUnregisterOnUnmount": false,
      "form": "Contracts",
      "getFormState": [Function],
      "initialValues": undefined,
      "keepDirtyOnReinitialize": false,
      "persistentSubmitErrors": false,
      "pure": true,
      "shouldAsyncValidate": [Function],
      "shouldValidate": [Function],
      "touchOnBlur": true,
      "touchOnChange": false,
    },
    "ref": "wrapped",
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "chainIds": Array [
          Object {
            "id": "0293f9b10a4453667db7fcfe74728c9d821add4b",
            "label": "new",
          },
        ],
        "changeContractFilter": [Function],
        "contracts": Object {
          "Cloner": Array [
            Object {
              "address": "d07e932212f7f368b6948ffd96e1d4c726c8395d",
              "createdAt": 1512480630000,
            },
            Object {
              "address": "2c6619f0418c2f191e2225091f7692363a91c336",
              "createdAt": 1512480770000,
            },
          ],
          "GreeterA": Array [
            Object {
              "address": "0293f9b10a4453667db7fcfe74728c9d821add4b",
              "createdAt": 1512481078000,
            },
          ],
          "GreeterB": Array [
            Object {
              "address": "GreeterA",
              "createdAt": 1512481078000,
            },
          ],
        },
        "destroyOnUnmount": true,
        "enableReinitialize": false,
        "fetchContracts": [Function],
        "filter": "Greeter",
        "forceUnregisterOnUnmount": false,
        "form": "Contracts",
        "getFormState": [Function],
        "initialValues": undefined,
        "keepDirtyOnReinitialize": false,
        "persistentSubmitErrors": false,
        "pure": true,
        "shouldAsyncValidate": [Function],
        "shouldValidate": [Function],
        "touchOnBlur": true,
        "touchOnChange": false,
      },
      "ref": "wrapped",
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenAdapter {
      "options": Object {
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
  },
}
`;

exports[`Contracts: index search contract input 1`] = `
Array [
  Array [
    "",
  ],
  Array [
    "UPDATE",
  ],
]
`;
